python/ray/data/_internal/compute.py

```diff
@@ -254,15 +254,25 @@ def _apply(
 
         if name is None:
             name = "map"
-        blocks_in = block_list.get_blocks_with_metadata()
-        # Bin blocks by target block size.
+        blocks_in: List[
+            Tuple[ObjectRef[Block], BlockMetadata]
+        ] = block_list.get_blocks_with_metadata()
+
+        # We bin blocks according to the following rules:
+        # 1. Attempt to bin up to the target block size.
+        # 2. If the max concurrency of the ActorPool is set, then
+        #    cap the number of bundles to match the size of the ActorPool.
+        #    This avoids additional overhead in submitting new actor tasks and allows
+        #    the actor task to do optimizations such as batch prefetching.
+        total_size = sum(metadata.num_rows for _, metadata in blocks_in)
         if target_block_size is not None:
             _check_batch_size(blocks_in, target_block_size, name)
-            block_bundles = _bundle_blocks_up_to_size(
-                blocks_in, target_block_size, name
-            )
-        else:
-            block_bundles = [((b,), (m,)) for b, m in blocks_in]
+        target_num_bundles = min(self.max_size, total_size / target_block_size)
+        target_block_size = total_size // target_num_bundles
```

Comment:
Nit: Can just do the min/capping on the target block size:
```suggestion
        target_block_size = min(target_block_size, total_size // self.max_size)
```